File: python/G.py
n, m = map(int, input().split())
G = {i+1:[] for i in range(n)}

for _ in range(m):
    x, y = map(int, input().split())
    G[x].append(y)
    
dp = [-1] * (n+1)


def rec(x):
    if dp[x] != -1:
        return dp[x]
    
    dp[x] = 0
    
    # 深ぼる
    for node in G[x]:
        dp[x] = max(dp[x], rec(node)+1)

    return dp[x]

for i in range(n):
    rec(i+1)
    
print(max(dp))
            
    
        

File: ruby/G.rb
n, m = gets.split.map(&:to_i)
@graph = Hash.new{|h, k| h[k] = []}
m.times do
  x, y = gets.split.map(&:to_i)
  @graph[x] << y
end

# ノードの名前が1からはじまるので、インデックス0はつかわない
@dp = [-1] * (n+1)

def rec(index)
  if @dp[index] != -1
    return @dp[index]
  end

  @dp[index] = 0

  @graph[index].each do |node|
    @dp[index] = [@dp[index], rec(node)+1].max
  end

  @dp[index]
end

# 各ノードを始点として探索する
(1..n).each do |index|
  rec(index)
end

print @dp.max